cardano-ledger-test/src/Test/Cardano/Ledger/Properties.hs

```diff
@@ -0,0 +1,794 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Test.Cardano.Ledger.Properties where
+
+import Cardano.Ledger.Alonzo (AlonzoEra, Value)
+import Cardano.Ledger.Alonzo.Data (Data, DataHash, hashData)
+import Cardano.Ledger.Alonzo.Language (Language (..))
+import Cardano.Ledger.Alonzo.PParams (PParams, PParams' (..))
+import Cardano.Ledger.Alonzo.Rules.Ledger (AlonzoLEDGER)
+import Cardano.Ledger.Alonzo.Rules.Utxow (AlonzoUTXOW)
+import Cardano.Ledger.Alonzo.Scripts
+import Cardano.Ledger.Alonzo.Tx
+  ( IsValid (..),
+    ValidatedTx (..),
+    hashScriptIntegrity,
+    minfee,
+  )
+import Cardano.Ledger.Alonzo.TxBody (TxBody (..), TxOut (..))
+import Cardano.Ledger.Alonzo.TxWitness
+  ( RdmrPtr (..),
+    Redeemers (..),
+    TxDats (..),
+    TxWitness (..),
+  )
+import Cardano.Ledger.BaseTypes
+  ( Network (..),
+    StrictMaybe (..),
+    strictMaybeToMaybe,
+  )
+import Cardano.Ledger.Coin (Coin (..))
+import Cardano.Ledger.Hashes (EraIndependentTxBody, ScriptHash (..))
+import Cardano.Ledger.Keys
+  ( GenDelegs (..),
+    KeyHash (..),
+    KeyPair (..),
+    KeyRole (..),
+    coerceKeyRole,
+    hashKey,
+  )
+import Cardano.Ledger.SafeHash (SafeHash, hashAnnotated)
+import Cardano.Ledger.ShelleyMA.Timelocks (Timelock (..), ValidityInterval (..))
+import Cardano.Ledger.Val
+import Cardano.Slotting.Slot (SlotNo (..))
+import Control.Monad (forM, join, replicateM)
+import Control.Monad.State.Strict (MonadState (..), modify)
+import Control.Monad.Trans.Class (MonadTrans (lift))
+import Control.Monad.Trans.RWS.Strict (RWST (..), ask)
+import Control.State.Transition.Extended hiding (Assertion)
+import Data.Bifunctor (first)
+import Data.Coerce
+import Data.Default.Class (Default (def))
+import qualified Data.Foldable as F
+import Data.Functor
+import Data.Map (Map)
+import qualified Data.Map as Map
+import Data.Maybe (catMaybes, fromJust)
+import Data.Monoid (All (..))
+import Data.Ratio ((%))
+import qualified Data.Sequence.Strict as Seq
+import qualified Data.Set as Set
+import GHC.Stack
+import Numeric.Natural
+import Plutus.V1.Ledger.Api (defaultCostModelParams)
+import Shelley.Spec.Ledger.API
+  ( Addr (..),
+    Credential (..),
+    LedgerEnv (LedgerEnv),
+    RewardAcnt (..),
+    StakeReference (..),
+    TxIn (..),
+    UTxO (..),
+    Wdrl (..),
+  )
+import Shelley.Spec.Ledger.EpochBoundary (obligation)
+import Shelley.Spec.Ledger.LedgerState
+  ( AccountState (..),
+    DPState (..),
+    DState (..),
+    PState (..),
+    RewardAccounts,
+    UTxOState (..),
+  )
+import Shelley.Spec.Ledger.STS.Utxo (UtxoEnv (..))
+import Shelley.Spec.Ledger.Tx (hashScript)
+import Shelley.Spec.Ledger.TxBody (DCert (..), DelegCert (..), Delegation (..), PoolParams (..))
+import Shelley.Spec.Ledger.UTxO (balance, makeWitnessVKey)
+import Test.Cardano.Ledger.Alonzo.Serialisation.Generators ()
```

Comment:
I guess I'm a bit worried when I see the serialisation generators imported in the real generators, since they're intended to be completely "dumb", whereas I can believe that in this module, we actually rely on certain properties of them.

Reply:
You bring up a great point. We must not have multiple instances of the same type for the same class in the same codebase, this is a bug. So, if generators in serialization are dumb, they need to be fixed to be smarter. In fact it would be wise to aggregate `Arbitrary` instances in one place per each era. 

I can certainly work on that in a separate PR.